Add actions to build Helm charts to the operators and retire os package building.

soenkeliebau opened the pull request:
fixes #4 
fixes https://github.com/stackabletech/issues/issues/135

Tested with:
- [x] Druid
- [x] Superset
- [x] Hive
- [x] Trino
- [x] Zookeeper
- [x] Regorule
- [x] Opa
- [x] Kafka
- [x] Nifi
- [x] Spark
- [x] HDFS
- [x] HBase

teozkr commented on repositories.yaml:
```diff
@@ -7,11 +7,11 @@ repositories:
-#     url: stackabletech/zookeeper-operator.git
-#     product_string: zookeeper
-#     pretty_string: Apache ZooKeeper
+   zookeeper-operator is disabled until the Helm chart is incorporated into the template.
```
```suggestion
```

teozkr commented on repositories.yaml:
```diff
@@ -54,4 +54,5 @@ repositories:
 # This is uncommented as I had issues with everything being deleted when this was just present as an empty key.
 # May be something to investigate.
 
```
```suggestion
```

teozkr commented on template/.github/workflows/publish_dev_artifacts.yml.j2:
```diff
@@ -18,62 +18,45 @@ env:
   REPO_APT_DEV_URL: https://repo.stackable.tech/repository/deb-dev
   REPO_RPM_DEV_URL: https://repo.stackable.tech/repository/rpm-dev
+  REPO_HELM_DEV_URL: https://repo.stackable.tech/repository/helm-dev
+  CT_CONFIG: deploy/helm/ct.yaml
```
```suggestion
  CHART_TESTING_CONFIG: deploy/helm/chart-testing.yaml
```

teozkr commented on template/.github/workflows/publish_dev_artifacts.yml.j2:
```diff
@@ -18,62 +18,45 @@ env:
   RUSTFLAGS: "-D warnings -W rust-2021-compatibility"
   REPO_APT_DEV_URL: https://repo.stackable.tech/repository/deb-dev
   REPO_RPM_DEV_URL: https://repo.stackable.tech/repository/rpm-dev
```
```suggestion
```

razvan replied:
Delete ?
```suggestion
```

teozkr commented on template/.github/workflows/publish_release_artifacts.yml.j2:
```diff
@@ -17,62 +17,35 @@ env:
   RUSTFLAGS: "-D warnings -W rust-2021-compatibility"
```
Same comments as on `publish_dev_artifacts.yml.j2`

razvan commented on template/.github/workflows/publish_dev_artifacts.yml.j2:
```diff
@@ -18,62 +18,45 @@ env:
+          version: v3.6.2
+      - name: Compile chart
+        run: make compile-chart
+      - name: Set up chart-testing
+        uses: helm/chart-testing-action@v2.1.0
+      # Disabled for now until we come up with a way
+      #- name: Run chart-testing (lint)
+      #  run: ct lint --config ${CT_CONFIG}
```
Delete ?
```suggestion
```

razvan commented on template/.github/workflows/publish_release_artifacts.yml.j2:
```diff
@@ -17,62 +17,35 @@ env:
   RUSTFLAGS: "-D warnings -W rust-2021-compatibility"
   REPO_APT_RELEASE_URL: https://repo.stackable.tech/repository/deb-release
   REPO_RPM_RELEASE_URL: https://repo.stackable.tech/repository/rpm-release
```
Delete ?
```suggestion
```

razvan commented on template/.github/workflows/publish_release_artifacts.yml.j2:
```diff
@@ -17,62 +17,35 @@ env:
+          version: v3.6.2
+      - name: Compile chart
+        run: make compile-chart
+      - name: Set up chart-testing
+        uses: helm/chart-testing-action@v2.1.0
+      - name: Run chart-testing (lint)
+        run: ct lint --config ${CT_CONFIG}
```
Delete or at least make the same as publish_dev_artifacts ?
```suggestion
```

teozkr commented on template/Makefile.j2:
```diff
@@ -0,0 +1,50 @@
+	rm -rf deploy/helm/{[ operator.product_string }]-operator/templates/crds.yaml
+
+version:
+	sed -i deploy/helm/{[ operator.product_string }]-operator/Chart.yaml "<version_placeholder>" ${VERSION}
```
```suggestion
        yq --in-place '.version = $ENV.VERSION | .appVersion = $ENV.VERSION' deploy/helm/{[ operator.product_string }]-operator/Chart.yaml
```

The current version fails silently if the command has already been executed.

teozkr commented on template/Makefile.j2:
```diff
@@ -0,0 +1,50 @@
+deploy/helm/{[ operator.product_string }]-operator/configs:
+	cp -r deploy/config-spec deploy/helm/{[ operator.product_string }]-operator/configs
+
+crds: deploy/helm/{[ operator.product_string }]-operator/templates/crds.yaml
```
CRDs should generally be in `$CHART_HOME/crds/`, which ensures that they are installed before the chart itself, and that they are not deleted even if the operator chart is.

hikhvar replied:
If you place CRDS in the `$CHART_HOME/crds/` directory, helm will not update them. You have to update them manually. To prevent the deletion on helm chart uninstall, I added the helm hook: `resource-policy=keep`:
https://github.com/stackabletech/operator-templating/pull/7/files#diff-9e6f51403e0a072f731bafd87e9c7d4ed0dca19e57cd445df0aaa31dc7d1dac8R36

hikhvar replied:
The operator itself does not need the charts before you can install it.

hikhvar replied:
https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#some-caveats-and-explanations

teozkr replied:
> The operator itself does not need the charts before you can install it.

That's true, but it prevents the same Helm Release from defining an object of that resource. That becomes an issue if any users try to pull in our operator as a Helm dependency.

> To prevent the deletion on helm chart uninstall, I added the helm hook: resource-policy=keep:

That's just delaying the issue though, since that means that reinstalling the chart will fail with a conflict error.

hikhvar replied:
So we have to decide. Either the helm chart can not update CRDs (putting them in the `crds` directory), or a user can't specify the helm-operator chart as depedency, if they install stackable resources in their helm chart. I think the first usecase (update CRDs) is much more common than the second usecase.

teozkr commented on template/Makefile.j2:
```diff
@@ -0,0 +1,50 @@
+
+kind:
+	kind create cluster --name chart-test || exit 0
+	kubectl --context kind-chart-test config view --minify --flatten > /tmp/kind.yaml
+	sleep 10
+	kubectl --kubeconfig /tmp/kind.yaml config set-cluster kind-chart-test --server=https://$(shell docker inspect chart-test-control-plane | jq '.[].NetworkSettings.Networks.kind.IPAddress' -r):6443
```
Not sure what you're trying to do here.. isn't this generally covered by `kind export kubeconfig --kubeconfig /tmp/kind.yaml`?

razvan commented on template/deploy/helm/[[product]]-operator/README.md.j2:
````diff
@@ -0,0 +1,46 @@
+## Install the Stackble Operator for {[ operator.pretty_string }]
+
+```bash
+helm install {[ operator.product_string }]-operator ./{[ operator.product_string }]-operator
````
This doesn't work if executed in the folder where the README is located.

I think this README file is misplaced and now that I see the folder structure I think it's unnecesserily complicated. Why have another folder (`product-operator`) under the `helm` folder ?

razvan commented on template/deploy/helm/[[product]]-operator/README.md.j2:
````diff
@@ -0,0 +1,46 @@
+```
+
+
+## Test the installed operator
+
+```bash
+helm test {[ operator.product_string }]-operator
+```
+
+This Helm test will create a {[ operator.product_string }] cluster and launch a Pod with kubectl that describes and deletes this cluster.
````
Delete ?

```suggestion
```

razvan reviewed (commented):
Needs work.

teozkr commented on template/deploy/helm/[[product]]-operator/templates/serviceaccount.yaml.j2:
```diff
@@ -0,0 +1,75 @@
+# =============
```
Looks like this is still "just" a regular gotpl?

teozkr commented on template/deploy/helm/[[product]]-operator/values.yaml.j2:
```diff
@@ -0,0 +1,74 @@
+  repository: docker.stackable.tech/stackable/{[ operator.product_string }]-operator
+  pullPolicy: IfNotPresent
+  # Overrides the image tag whose default is the chart appVersion.
+  tag: "latest"
```
This is really dangerous to specify in `values.yaml`, since it'll be kept during upgrades even if the chart itself changed.

hikhvar replied:
We should undefine this. In the template, we set it to "appVersion" of we leave `tag` empty.

siegfriedweber commented on template/deploy/helm/[[product]]-operator/templates/serviceaccount.yaml.j2:
```diff
@@ -0,0 +1,75 @@
+# This file is automatically generated from the templates in stackabletech/operator-templating
+# DON'T MANUALLY EDIT THIS FILE
+# =============
+{{- if .Values.serviceAccount.create -}}
```
```suggestion
{{ if .Values.serviceAccount.create -}}
```

Otherwise the rendered template looks like this:
```
# This file is automatically generated from the templates in stackabletech/operator-templating
# DON'T MANUALLY EDIT THIS FILE
# =============apiVersion: v1
kind: ServiceAccount
```

siegfriedweber commented on template/deploy/helm/[[product]]-operator/templates/serviceaccount.yaml.j2:
```diff
@@ -0,0 +1,75 @@
+  name: {{ include "operator.serviceAccountName" . }}-role
+  labels:
+    {{- include "operator.labels" . | nindent 4 }}
+rules:
```
The rules are operator-specific and should be configurable.

see [rules for the Superset operator](https://github.com/stackabletech/superset-operator/blob/4892ca21162ccf187ace4b5b276fb17a9f1a2a06/deploy/role/superset-cluster-role.yaml)

siegfriedweber commented on template/deploy/helm/[[product]]-operator/values.yaml.j2:
```diff
@@ -0,0 +1,74 @@
+  # runAsNonRoot: true
+  # runAsUser: 1000
+
+service:
+  type: ClusterIP
+  port: 80
+
+ingress:
+  enabled: false
+  annotations: {}
+    # kubernetes.io/ingress.class: nginx
+    # kubernetes.io/tls-acme: "true"
+  hosts:
+    - host: chart-example.local
+      paths: []
+  tls: []
+  #  - secretName: chart-example-tls
+  #    hosts:
+  #      - chart-example.local
```
Unused. Remove from Helm values or create according templates.

soenkeliebau commented:
I've discussed this PR with @siegfriedweber  and we agreed that the best way forward would be to include parts in the template, but also allowing to add parts directly in the operator repositories.

I've addressed the comments as best I could. Sigi will have a look at the Hive operator and do some additional testing before we merge this.

maltesander reviewed (approved):
Tested and works. LGTM.
